xdcs-server/src/main/java/pl/edu/agh/xdcs/agentapi/impl/or/UploadObjectsStreamObserver.java

```diff
@@ -0,0 +1,91 @@
+package pl.edu.agh.xdcs.agentapi.impl.or;
+
+import io.grpc.stub.StreamObserver;
+import pl.edu.agh.xdcs.agentapi.utils.GrpcServiceException;
+import pl.edu.agh.xdcs.api.Chunk;
+import pl.edu.agh.xdcs.api.OkResponse;
+import pl.edu.agh.xdcs.or.ObjectRepository;
+
+import javax.enterprise.concurrent.ManagedThreadFactory;
+import java.io.IOException;
+import java.io.PipedInputStream;
+import java.io.PipedOutputStream;
+import java.io.UncheckedIOException;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+/**
+ * @author Kamil Jarosz
+ */
+public class UploadObjectsStreamObserver implements StreamObserver<Chunk> {
+    private final StreamObserver<OkResponse> responseObserver;
+    private final ObjectRepository objectRepository;
+    private final ZipInputStream inputStream;
+    private final PipedOutputStream pipedOutput;
+
+    UploadObjectsStreamObserver(
+            ManagedThreadFactory threadFactory,
+            ObjectRepository objectRepository,
+            StreamObserver<OkResponse> responseObserver) {
+        this.objectRepository = objectRepository;
+        this.responseObserver = responseObserver;
+        this.pipedOutput = new PipedOutputStream();
+        try {
+            this.inputStream = new ZipInputStream(new PipedInputStream(pipedOutput));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+
+        threadFactory.newThread(() -> {
+            try {
+                receiveAsync();
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }).start();
+    }
+
+    private void receiveAsync() throws IOException {
+        try {
+            while (true) {
+                ZipEntry nextEntry = inputStream.getNextEntry();
+                if (nextEntry == null) return;
+
+                String objectId = nextEntry.getName();
+                String storedObjectId = objectRepository.store(inputStream);
+
+                if (!storedObjectId.equals(objectId)) {
+                    responseObserver.onError(new GrpcServiceException(
+                            "Object IDs don't match: " + objectId + ", " + storedObjectId));
```

Comment:
w przypadku tego błędu zostają śmieci w OR, możnaby to wyczyścić

Reply:
właśnie nie za bardzo można, nie wiesz czy ktoś inny nie używa tego obiektu